### PR TITLE
Implement `bal test --cloud` and execution of the created docker containers

### PR DESCRIPTION
**Pull request by Ramith-D-Rodrigo**
## Purpose
Similar to `bal build --cloud`, executing ballerina tests in cloud containers by implementing the `bal test --cloud` command. 
Resolves the issue #695 

## Goals
Encapsulating the test execution dependencies jars and the test suite json in a container to execute the bal test command.
Create a separate fat jar and use that for building the docker container and executing it.

~~Currently this is achieved for the JVM path only.~~
GraalVM execution path is also added.

As discussed, the unit tests will be made after the initial code review.

UPDATE: 
All the relevant tests have been added.

## Approach
The lifecycle listener plugin notifies C2C after emitting the test executable fat jar with the artifact type (BUILD or TEST). Based on that type, test flow is implemented. In the implementation, the test suite json is modified, relevant command line arguments are added, test execution dependencies are copied, and dockerfile is generated accordingly.

Once generated, the docker command is executed using the ProcessBuilder.

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

**Review by anuruddhal — CHANGES_REQUESTED** [review]
compiler-plugin/src/main/java/io/ballerina/c2c/models/KubernetesDataHolder.java is an empty file.

**Comment by Ramith-D-Rodrigo** [comment]
> compiler-plugin/src/main/java/io/ballerina/c2c/models/KubernetesDataHolder.java is an empty file.

Not sure why but viewing the file shows the content and it is not empty.

**Comment by Ramith-D-Rodrigo** [comment]
Currently testing the changes using the Full Build Pipeline.

File: compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-code-coverage/tests/some_tests.bal [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+@test:Config{}
+function anotherSmallTest() {
+    test:assertEquals(2, 2, "2 should be equal to 2");
+}
+
+@test:Config{}
+function anotherSmallTest2() {
+    test:assertNotEquals(2, 3, "2 should not be equal to 3");
+}
```
Comment (anuruddhal) [inline compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-code-coverage/tests/some_tests.bal:32]: Add new line

File: compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-graalvm-mocking/modules/mod1/mod1.bal [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+function intMult(int a, int b) returns int {
+    return a * b;
+}
```
Comment (anuruddhal) [inline compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-graalvm-mocking/modules/mod1/mod1.bal:19]: Add new line

File: compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-graalvm-mocking/modules/mod1/tests/lib_test.bal [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+test:MockFunction intMultMockFn = new ();
+
+@test:Config{}
+function testMockingInModule() {
+    test:when(intMultMockFn).thenReturn(33);
+    test:assertEquals(intMult(1, 2), 33, "Mocking failed");
+
+    test:when(intMultMockFn).withArguments(6,6).callOriginal();
+    test:assertEquals(intMult(6, 6), 36, "Mocking failed");
+}
```
Comment (anuruddhal) [inline compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-graalvm-mocking/modules/mod1/tests/lib_test.bal:29]: Add new line

File: compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-graalvm/tests/tests_without_mocks.bal [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/test;
+
+@test:Config{}
+function assertThis() {
+    test:assertEquals(1, 1, "1 should be equal to 1");
+}
```
Comment (anuruddhal) [inline compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-graalvm/tests/tests_without_mocks.bal:22]: Add new line

File: compiler-plugin/src/main/java/io/ballerina/c2c/utils/DockerGenerator.java [hunk 5]
```diff
@@ -227,12 +261,122 @@ private String entryPointArgBuilder(List<String> args) {
...
+        new TreeSet<>(this.dockerModel.getDependencyJarPaths())
+                .stream()
+                .map(Path::getFileName)
+                .forEach(path -> {
+                    testDockerFileContent.append("COPY ")
+                                    .append(path)
+                                    .append(" ").append(getWorkDir())
+                                    .append("/jars/ ").append(LINE_SEPARATOR);
+                            //TODO: Remove once https://github.com/moby/moby/issues/37965 is fixed.
+                            boolean isCiBuild = "true".equals(System.getenv().get("CI_BUILD"));
```
Comment (anuruddhal) [inline compiler-plugin/src/main/java/io/ballerina/c2c/utils/DockerGenerator.java:293]: We can remove this `CI_BUILD` configs now

**Review by anuruddhal — CHANGES_REQUESTED** [review]

File: compiler-plugin/src/main/java/io/ballerina/c2c/utils/DockerGenerator.java (hunk 5 above)
Comment (Ramith-D-Rodrigo) [reply compiler-plugin/src/main/java/io/ballerina/c2c/utils/DockerGenerator.java:293]: > We can remove this `CI_BUILD` configs now

Shall I remove from both places? Line no:206 and 292

File: compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-code-coverage/tests/some_tests.bal (hunk 1 above)
Comment (Ramith-D-Rodrigo) [reply compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-code-coverage/tests/some_tests.bal]: Added in [5fc44be](https://github.com/ballerina-platform/module-ballerina-c2c/pull/765/commits/5fc44befe4d898c702ef885cea29fd93df2f73cb)

File: compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-graalvm-mocking/modules/mod1/mod1.bal (hunk 2 above)
Comment (Ramith-D-Rodrigo) [reply compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-graalvm-mocking/modules/mod1/mod1.bal]: Added in [5fc44be](https://github.com/ballerina-platform/module-ballerina-c2c/pull/765/commits/5fc44befe4d898c702ef885cea29fd93df2f73cb)

File: compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-graalvm-mocking/modules/mod1/tests/lib_test.bal (hunk 3 above)
Comment (Ramith-D-Rodrigo) [reply compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-graalvm-mocking/modules/mod1/tests/lib_test.bal]: Added in [5fc44be](https://github.com/ballerina-platform/module-ballerina-c2c/pull/765/commits/5fc44befe4d898c702ef885cea29fd93df2f73cb)

File: compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-graalvm/tests/tests_without_mocks.bal (hunk 4 above)
Comment (Ramith-D-Rodrigo) [reply compiler-plugin-tests/src/test/resources/docker-cloud-test/project-based-tests/cloud-flag-graalvm/tests/tests_without_mocks.bal]: Added in [5fc44be](https://github.com/ballerina-platform/module-ballerina-c2c/pull/765/commits/5fc44befe4d898c702ef885cea29fd93df2f73cb)

File: compiler-plugin/src/main/java/io/ballerina/c2c/utils/DockerGenerator.java (hunk 5 above)
Comment (Ramith-D-Rodrigo) [reply compiler-plugin/src/main/java/io/ballerina/c2c/utils/DockerGenerator.java]: Done in [5fc44be](https://github.com/ballerina-platform/module-ballerina-c2c/pull/765/commits/5fc44befe4d898c702ef885cea29fd93df2f73cb)

File: compiler-plugin/src/main/java/io/ballerina/c2c/tasks/C2CCodeGeneratedTask.java [hunk 6]
```diff
@@ -69,25 +99,147 @@ public void perform(CompilerLifecycleEventContext compilerLifecycleEventContext)
...
+                String outputName = "$anon".equals(descriptor.org().value()) ? extractJarName(path.getFileName()) :
+                        descriptor.name().value().toLowerCase(Locale.ROOT);
+                dataHolder.setOutputName(outputName);
+                addDependencyJars(compilerLifecycleEventContext.compilation(), executableJarName);
+                dataHolder.setSourceRoot(executablePath.get().getParent()
+                        .getParent().getParent());
+                codeGeneratedInternal(KubernetesUtils.getProjectID(currentPackage),
+                        path, compilerLifecycleEventContext.currentPackage(), balCommand);
+            });
+        } else {
```
Comment (xlight05) [inline compiler-plugin/src/main/java/io/ballerina/c2c/tasks/C2CCodeGeneratedTask.java:122]: We can explicitly check for BalCommand.TEST here

File: compiler-plugin/src/main/java/io/ballerina/c2c/tasks/C2CCodeGeneratedTask.java [hunk 7]
```diff
@@ -69,25 +99,147 @@ public void perform(CompilerLifecycleEventContext compilerLifecycleEventContext)
...
-            String outputName = "$anon".equals(descriptor.org().value()) ? extractJarName(path.getFileName()) :
-                    descriptor.name().value().toLowerCase(Locale.ROOT);
-            dataHolder.setOutputName(outputName);
-            addDependencyJars(compilerLifecycleEventContext.compilation(), executableJarName);
-            dataHolder.setSourceRoot(executablePath.get().getParent()
-                    .getParent().getParent());
-            codeGeneratedInternal(KubernetesUtils.getProjectID(currentPackage),
-                    path, compilerLifecycleEventContext.currentPackage());
+
+        if (balCommand == BalCommand.BUILD) {
```
Comment (xlight05) [inline compiler-plugin/src/main/java/io/ballerina/c2c/tasks/C2CCodeGeneratedTask.java:107]: Shall we move these build processing and test processing in to separate methods?

Comment (Ramith-D-Rodrigo) [reply compiler-plugin/src/main/java/io/ballerina/c2c/tasks/C2CCodeGeneratedTask.java:107]: Done with [e332a9e](https://github.com/ballerina-platform/module-ballerina-c2c/pull/765/commits/e332a9ea20271f7c481c3750bad97af0f5f32df1)

File: compiler-plugin/src/main/java/io/ballerina/c2c/tasks/C2CCodeGeneratedTask.java (hunk 6 above)
Comment (Ramith-D-Rodrigo) [reply compiler-plugin/src/main/java/io/ballerina/c2c/tasks/C2CCodeGeneratedTask.java]: Added in [e332a9e](https://github.com/ballerina-platform/module-ballerina-c2c/pull/765/commits/e332a9ea20271f7c481c3750bad97af0f5f32df1)

**Review by xlight05 — APPROVED** [review]

**Pull request merged**